Add Cudd as an alternative for miniBDD

romainbrenguier opened the pull request:
This PR allows to optionally download and compile with CUDD instead of miniBDD.
This is in preparation for https://github.com/diffblue/cbmc/pull/3730

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [na] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [na] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

peterschrammel commented on src/util/CMakeLists.txt:
```diff
@@ -1,5 +1,6 @@
 file(GLOB_RECURSE sources "*.cpp" "*.h")
 
+
```
spurious change

peterschrammel commented on src/config.inc:
```diff
@@ -72,3 +72,6 @@ OSX_IDENTITY="Developer ID Application: Daniel Kroening"
 # Detailed version information
 CBMC_VERSION = 5.11
+
+# Use the CUDD library for BDDs, can be installed using `make -C src cudd-download`
```
maybe needs documentation in COMPILING.md as well

martin-cs replied:
Yes please.

romainbrenguier replied:
added

martin-cs reviewed (approved):
As a general point, is there enough happening in the BDD space that we need to support others or is it pretty much over and there is only one choice?

romainbrenguier commented:
@martin-cs 
> As a general point, is there enough happening in the BDD space that we need to support others or is it pretty much over and there is only one choice?

I'm not planning to add support for others soon, but it could a possibility in the long term. @forejtv was keen to be able to easily switch to another library if ever needed.

martin-cs commented:
On Tue, 2019-01-29 at 07:48 +0000, Romain Brenguier wrote:
> @martin-cs 
> > As a general point, is there enough happening in the BDD space that
> > we need to support others or is it pretty much over and there is
> > only one choice?
> 
> I'm not planning to add support for others soon, but it could a
> possibility in the long term. @forejtv was keen to be able to easily
> switch to another library if ever needed.

I guess my question was more along the lines of "is there another
library?".  Last time I checked BDD development wasn't the most lively
of research areas.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: c47dbf71eeec4abbf713a96ce78d36c7b9d00020).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99238970

allredj commented on COMPILING.md:
```diff
@@ -282,3 +282,30 @@ To work with Eclipse, do the following:
+
+## Compiling with CUDD
+
+Cudd is BDD library which can be used instead of the builtin miniBDD
```
`BDD` -> `a BDD`

allredj commented on COMPILING.md:
```diff
@@ -282,3 +282,30 @@ To work with Eclipse, do the following:
+
+If compiling with make:
+
+1. Download and complie Cudd:
```
`complie` -> `compile`

allredj reviewed (changes requested):
Only looked at last commit.
Some typos. Otherwise instructions seem to work.

romainbrenguier commented:
@allredj I fixed the typos

allredj reviewed (approved):
Only reviewed commit "Instructions for compiling with Cudd"

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9d5e1b1cda8c05f90be156af9d1dd58e9f2ce353).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99262268

romainbrenguier commented:
@tautschnig can you have a look at this?

tautschnig commented on scripts/cudd-3.0.0-patch:
```diff
@@ -0,0 +1,12 @@
+ #include <string>
+ #include <vector>
+-#include "cudd.h"
++#include "../cudd/cudd.h"
```
Why is this kind of patching necessary, can't we just add the appropriate include path to compilation commands (like we do for SAT solvers)?

tautschnig commented on src/Makefile:
```diff
@@ -109,6 +109,17 @@ minisat2-download:
+	@echo "Downloading Cudd 3.0.0"
+	@$(DOWNLOADER) https://sourceforge.net/projects/cudd-mirror/files/cudd-3.0.0.tar.gz/download cudd-3.0.0.tar.gz
+	@$(TAR) xfz cudd-3.0.0.tar.gz
+	@rm -Rf ../cudd-3.0.0
```
Nit pick: use `@$(RM) -r ../cudd-3.0.0`

tautschnig commented on src/Makefile:
```diff
@@ -109,6 +109,17 @@ minisat2-download:
+	@$(TAR) xfz cudd-3.0.0.tar.gz
+	@rm -Rf ../cudd-3.0.0
+	@mv cudd-3.0.0 ../
+	@rm cudd-3.0.0.tar.gz
```
`@$(RM) cudd-3.0.0.tar.gz`

tautschnig commented on src/Makefile:
```diff
@@ -109,6 +109,17 @@ minisat2-download:
+	@rm cudd-3.0.0.tar.gz
+	@(cd ../cudd-3.0.0; patch -p1 < ../scripts/cudd-3.0.0-patch)
+	@echo "Compiling Cudd 3.0.0"
+	@(cd ../cudd-3.0.0; ./configure; make)
```
Use `$(MAKE)` instead of `make`

tautschnig commented on .travis.yml:
```diff
@@ -253,10 +253,12 @@ jobs:
         - ccache -z
         - ccache --max-size=1G
-        - cmake -H. -Bbuild '-DCMAKE_BUILD_TYPE=Release'  '-DCMAKE_CXX_COMPILER=/usr/bin/g++-7'
+        - cmake -H. -Bbuild '-DCMAKE_BUILD_TYPE=Release'  '-DCMAKE_CXX_COMPILER=/usr/bin/g++-7' '-DCMAKE_USE_CUDD=true'
```
How much longer does it take to do this?

romainbrenguier replied:
On my computer the configuration takes a few second more and the compilation about 40sec.
On travis, this job took 18m52, and the job that came before and is kind of similar took 17m44.

tautschnig replied:
Great, thanks a lot, maybe you can put that info in the commit message?

tautschnig commented on .travis.yml:
```diff
@@ -253,10 +253,12 @@ jobs:
-      script: (cd build; ctest -V -L CORE -j2)
+      script:
+        - (cd build; ctest -V -L CORE -j2)
+        - (cd build; bin/unit -r xml -d yes --order lex "[core][solver]")
```
Would you mind explaining that one?

romainbrenguier replied:
I added a comment in the commit message. There are unit tests (in particular for bdd_expr) which depend on BDDs so they could be affect when we swap the implementation from miniBDD to cudd, that's why I run them in that step

tautschnig replied:
But wouldn't `ctest` run them anyway?

romainbrenguier replied:
Yes I think you are right, `unit` has the label `CORE` in the cmake file so it should be run, I will remove that step

tautschnig commented on src/solvers/bdd/bdd_cudd.h:
```diff
@@ -0,0 +1,158 @@
+    return Cudd_IsComplement(node) != 0;
+  }
+
+  unsigned int index() const
```
Can we also make this `std::size_t`?

tautschnig commented on src/solvers/bdd/bdd_cudd.h:
```diff
@@ -0,0 +1,158 @@
+    return bdd_nodet(Cudd_E(node));
+  }
+
+  long id() const
```
Why is it `long`? And could at least this function be documented? The others are all pretty intuitive, but this one isn't (that is, how is it different from `index`)?

tautschnig commented on src/solvers/bdd/bdd_cudd.h:
```diff
@@ -0,0 +1,158 @@
+class bddt
+{
+public:
+  bddt() = delete;
```
No need to, there is another constructor so this one won't be auto-generated.

tautschnig commented on src/solvers/bdd/bdd_cudd.h:
```diff
@@ -0,0 +1,158 @@
+    return bddt(cudd.bddZero());
+  }
+
+  bddt bdd_variable(unsigned int index)
```
`std::size_t`?

tautschnig commented on src/solvers/bdd/bdd_cudd.h:
```diff
@@ -0,0 +1,158 @@
+/// \file
+/// Interface to Cudd BDD functions that are used in CBMC
+/// To ensure we can easily switch from Cudd to other BDD libraries if needed,
+/// the BDD functions should only be accessed from this header file.
```
That doesn't quite seem to be the case, there is an unconditional include of cudd here.

tautschnig commented on src/solvers/bdd/bdd_miniBDD.h:
```diff
@@ -0,0 +1,173 @@
+/// \file
+/// Interface to miniBDD functions that are used in CBMC
+/// To ensure we can easily switch to other BDD libraries if needed,
+/// the BDD functions should only be accessed from this header file.
```
Sounds like the same comment as in the CUDD file, and I equally don't fully understand it. It seems that the functions in this file should _not_ be directly used?!

romainbrenguier replied:
I rephrased a bit this documentation, is it clearer?

tautschnig replied:
Yes, I think that's fine now.

tautschnig commented on src/solvers/bdd/bdd_miniBDD.h:
```diff
@@ -0,0 +1,173 @@
+#ifndef CPROVER_SOLVERS_BDD_BDD_MINIBDD_H
+#define CPROVER_SOLVERS_BDD_BDD_MINIBDD_H
+
+#include "unordered_map"
```
Angle brackets?

tautschnig commented on src/solvers/bdd/bdd_miniBDD.h:
```diff
@@ -0,0 +1,173 @@
+class bddt;
+class bdd_nodet;
+
+static std::unordered_map<unsigned int, unsigned int> bdd_var_to_index;
```
No, don't put a `static` declaration in a header file. You'll have N copies of this one, for each translation unit that ends up including this header file.

romainbrenguier replied:
Yes of course, thanks. Sorry I forgot to clean-up this hack.

tautschnig commented on src/solvers/bdd/bdd_miniBDD.h:
```diff
@@ -0,0 +1,173 @@
+    return node->node_number == 0;
+  }
+
+  unsigned int index() const
```
See CUDD: `std::size_t`?

tautschnig commented on src/solvers/bdd/bdd_miniBDD.h:
```diff
@@ -0,0 +1,173 @@
+    return bdd_nodet(node->low.node);
+  }
+
+  long id() const
```
See CUDD comments.

tautschnig commented on src/solvers/bdd/bdd_miniBDD.h:
```diff
@@ -0,0 +1,173 @@
+static std::unordered_map<unsigned int, unsigned int> bdd_var_to_index;
+
+/// Low level access to the structure of the BDD, read-only.
+class bdd_nodet
```
If we have multiple classes implementing the same interface, then there should be a single common interface declaration.

romainbrenguier replied:
Only one these classes should be compiled, (which one depends on the HAVE_CUDD flag). I want to avoid virtual functions.

tautschnig replied:
You could use [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern)? Not a show-stopper though, I'm just worried that over time these interfaces might get out-of-sync, any maybe they already are.

tautschnig commented on src/solvers/bdd/bdd_miniBDD.h:
```diff
@@ -0,0 +1,173 @@
+
+  bddt bdd_or(const bddt &other) const
+  {
+    return bddt((*this) | (other));
```
Not sure why the parentheses are required.

tautschnig commented on src/solvers/bdd/bdd_miniBDD.h:
```diff
@@ -0,0 +1,173 @@
+    return bddt(*this);
+  }
+
+  bddt &operator=(const bddt &other)
```
The CUDD interface doesn't have this.

romainbrenguier replied:
I think it would have a default copy which works fine. miniBDD on the other hand crashes if we try to  assign to BDDs which are equal.
I will add the operator to the cudd interface as well to make it clearer.

tautschnig commented on src/solvers/bdd/bdd_miniBDD.h:
```diff
@@ -0,0 +1,173 @@
+    return bddt((*this) ^ (other));
+  }
+
+  static bddt bdd_ite(const bddt &i, const bddt &t, const bddt &e)
```
Did I miss this in the CUDD header?

romainbrenguier replied:
I guess so

tautschnig commented on src/solvers/bdd/miniBDD/example.cpp:
```diff
@@ -20,46 +20,46 @@ int main()
   miniBDD::BDD result;
 
-  #if 0
+#if 0
```
Please keep moving and clang-formatting in separate commits.

romainbrenguier replied:
done

tautschnig reviewed (changes requested):
Mostly requesting small and/or administrative changes. And I'm unhappy about the absence of commit message bodies, but I guess I've said that before. For the sake of reviewing we would likely have been better off having the miniBDD related changes in a PR of it's own and then follow-up with CUDD. But now it's reviewed so that's just a lesson for the future.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: ab652d43d06389c3ef1de8c968d617f2eef718d7).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99302331
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 83097cda6bffcfac3752cf9b227ef4755fb6b483).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99317324
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 345c393fc6c6861d0570ed7d96ccfde3afd4264e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99319498

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 9c2f5e94bb53208a0e225dcc14a892e1b04b2a1c).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99332199

romainbrenguier commented:
@tautschnig I think I addressed all the comments. What would be left is using CRTP. That would be nice but I'm not sure it's worth it. Since we only have two BDD implementations and both are tested on CI there isn't too much risk, and I prefer to keep it simple.

tautschnig reviewed (approved):
Looks ok to me except "Move miniBDD to a new bdd module" still has a lot of whitespace changes mixed into the file move, which needs to be split out in two separate commits (a clean move and the clang-format changes either before or after the move).

romainbrenguier commented:
> Looks ok to me except "Move miniBDD to a new bdd module" still has a lot of whitespace changes mixed into the file move, which needs to be split out in two separate commits (a clean move and the clang-format changes either before or after the move).

I added clang-format commit before to make the move clean

romainbrenguier committed:
Reformatting miniBDD/example.cpp

This will get reformatted by clang-format while doing unrelated changes
if we don't take care of it before.

romainbrenguier committed:
Reformatting miniBDD.cpp

No functional change

romainbrenguier committed:
Reformatting miniBDD.h

No functional change

romainbrenguier committed:
Reformatting miniBDD unit test

romainbrenguier committed:
Move miniBDD to a new bdd module

This will allow us to have several implementations of BDD among which we
choose at compile time.

romainbrenguier committed:
Fix miniBDD example

There were mistakes in the file that prevented it from being compiled.

romainbrenguier committed:
Unifying BDD interface for miniBDD

Add a header file that will provide functions that are used in cbmc and
which could be substituted for another bdd library.

romainbrenguier committed:
Use unified BDD interface instead of miniBDD

We make the code use the unified interface to BDDs rather than miniBDD
directly so that we can switch to another implementation easily.

romainbrenguier committed:
Unit test for bdd_expr

Add a couple of basic tests for bdd_expr, in order to check that there
is no obvious mistake in the use of BDDs from bdd_expr.

romainbrenguier committed:
Adding Makefile instruction to install CUDD

Add a cudd-download entry in the Makefile to download and compile Cudd.
This is an optional step, but requiered if we want to compile with the
HAVE_CUDD flag.

romainbrenguier committed:
Add CUDD to CMake

CMake instructions to download Cudd when the CMAKE_USE_CUDD parameter is
set to true.

romainbrenguier committed:
Use Cudd in one of the travis build

This build will compile with CMAKE_USE_CUDD to check that compilation
with CUDD works correctly.

From testing it on travis, it seems this makes the build about 1 minute
longer.

romainbrenguier committed:
Interface to Cudd BDD functions

We define this interface to make it easy to switch from Cudd to another
library if it is ever needed.

romainbrenguier committed:
Header with the choice between cudd and miniBDD

This header will either include bdd_cudd or bdd_miniBDD depending on
whether the HAVE_CUDD flag is defined.

romainbrenguier committed:
Use emplace instead of insert

No functional change

romainbrenguier committed:
Makefile instruction to compile with CUDD

In case the CUDD variable, which should be pointing to the installation
directory of CUDD is defined, we link to it and define the HAVE_CUDD
flag to let the CBMC code know that it is available.

romainbrenguier committed:
Instructions for compiling with Cudd

Add explenation on how to compile CBMC so that it uses Cudd as the BDD implementation.

allredj reviewed (commented):
:no_entry_sign:
This PR failed Diffblue compatibility checks (cbmc commit: 06b403dfc4e044abf82db52172b0e9a4c6fb5b7e).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99385610
Status will be re-evaluated on next push.
Please contact @peterschrammel, @thk123, or @allredj for support.

Common spurious failures:
  - the cbmc commit has disappeared in the mean time (e.g. in a force-push)
  - the author is not in the list of contributors (e.g. first-time contributors).

The incompatibility may have been introduced by an earlier PR. In that case merging this
PR should be avoided unless it fixes the current incompatibility.

allredj reviewed (commented):
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 7dd895995fc3e2bdbd4a85e65aa5c817a9848a35).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/99385913
